### ISSUE TITLE
Vectorize Model.embed_phon_tokens: Python double loop costs 17-25% of every train step (68x CPU / 147x CUDA available)

### ISSUE THREAD

**Issue by mathnathan** [emelex-ai/BRIDGE#221]
## Summary

`Model.embed_phon_tokens` ([`bridge/domain/model/model.py:135`](https://github.com/emelex-ai/BRIDGE/blob/main/bridge/domain/model/model.py#L135)) averages phonological feature embeddings using a **Python double loop over every (batch, position) pair**, writing one row at a time into a preallocated tensor. It is the single largest CPU-side cost in both training and generation.

Replacing the loop with one padded gather + masked mean is **68× faster on CPU and 147× faster on CUDA** (forward-only), and **30×/45× including backward** — with bitwise-identical output on CPU.

The catch, and the reason this is an issue rather than a one-line patch: **the win only materialises if the padded index tensor is built by the tokenizer, not inside `embed_phon_tokens`.** Building the padding per call recovers only ~2×. See "Why the naive fix doesn't work" below.

## The current code

```python
def embed_phon_tokens(self, tokens) -> torch.Tensor:
    batch_size = len(tokens)
    max_phon_len = len(tokens[0])
    output_embedding = torch.zeros(
        (batch_size, max_phon_len, self.model_config.d_model), device=self.device
    )
    for batch_num, batch in enumerate(tokens):
        for indx, tokes in enumerate(batch):
            avg_embedding = self.phonology_embedding(tokes).mean(axis=0)
            output_embedding[batch_num, indx, :] = avg_embedding
    return output_embedding + self.phon_position_embedding.weight[None, : len(tokens[0])]
```

`tokens` is `list[list[Tensor]]` — a ragged structure where `tokens[b][i]` is a 1-D tensor of the *active phonetic feature indices* for phoneme `i` of batch item `b`. Each of the `B × L` inner iterations performs a separate embedding gather, a separate `mean`, and a separate indexed write — and creates a separate autograd node for the backward pass.

Call sites: `model.py:217` (`_decode_phon`), `model.py:239` (`embed_p`), `model.py:248` (`embed_op`), plus `model.py:523` and `model.py:626` inside the generation loops.

## Measured impact

Measured on this repo at commit `8285bff` + the domain-simplification changes, RTX 5080 / 12-core CPU, `torch.set_num_threads(1)`, batch size 32, `d_model=64`, corpus `tests/domain/model/data/data.csv`. Harness in the appendix — all numbers below were reproduced directly, not estimated.

### Isolated (B=32, L=7, F_max=7 → 224 Python iterations per call)

| | CPU | CUDA |
|---|---|---|
| forward, current | 6.90 ms | 10.64 ms |
| forward, vectorized (padding prebuilt) | 0.101 ms | 0.072 ms |
| **forward speedup** | **68.6×** | **146.8×** |
| fwd+bwd, current | 14.30 ms | 32.21 ms |
| fwd+bwd, vectorized | 0.469 ms | 0.712 ms |
| **fwd+bwd speedup** | **30.5×** | **45.3×** |

### Share of end-to-end work

| | CPU | CUDA |
|---|---|---|
| o2p train step (B=32) | 47.5 ms | 61.4 ms |
| ...of which `embed_phon_tokens` forward | 8.3 ms (**17%**) | 15.5 ms (**25%**) |
| Python iterations per train step | 262 | 262 |
| `generate()` o2p (B=32, deterministic) | 34 ms | 21 ms |
| ...of which `embed_phon_tokens` | 16 ms (**47%**) | 5 ms (**24%**) |

The forward-only share understates the true cost, since the `B × L` autograd nodes are also paid in the backward pass (compare the fwd+bwd rows).

> Note on the `generate()` row: with an untrained model, generation length is random (EOS is sampled at an arbitrary step), so the call count and total vary run to run. The structural point is stable — see below.

### Why generation is worse than the table suggests

`embed_phon_tokens` is called **once per decode step, over the entire growing prefix** (`model.py:626`), rather than embedding only the newly generated position. That makes generation `O(B · L²)` in Python iterations: a full 29-step run costs roughly `32 · 29² / 2 ≈ 13,500` iterations. Fixing the loop bound is a separate (also worthwhile) optimisation — see "Related" — but vectorizing the inner work removes the constant factor regardless.

## Why the naive fix doesn't work

The obvious patch — build the padded tensor inside `embed_phon_tokens` — was measured and is **not** worth doing on its own:

| variant | CPU | CUDA |
|---|---|---|
| current | 6.90 ms | 10.64 ms |
| vectorized, padding built **inside the call** | 2.65 ms (2.6×) | 5.63 ms (1.9×) |
| vectorized, padding **prebuilt** | 0.101 ms (68.6×) | 0.072 ms (146.8×) |

Building the `(B, L, F_max)` index tensor with `idx[b, i, :len(t)] = t` is itself a `B × L` Python loop doing per-element device writes — it just relocates the bottleneck. **The padded representation has to come from the tokenizer.**

### Part 1 — tokenizer emits a padded, batched representation

`PhonemeTokenizer.encode` ([`bridge/domain/tokenizer/phoneme_tokenizer.py:194`](https://github.com/emelex-ai/BRIDGE/blob/main/bridge/domain/tokenizer/phoneme_tokenizer.py#L194)) **already loops over exactly this data** to build `enc_indices` / `dec_indices` as `list[list[Tensor]]`. Have it produce, alongside (or instead of) the ragged lists:

- `feature_ids: LongTensor (B, L, F_max)` — active feature indices, zero-padded
- `feature_mask: BoolTensor (B, L, F_max)` — `True` where a slot holds a real feature

This is a change to the `EncodingComponent` contract (`bridge/domain/datamodels/encodings.py`), so it touches `BridgeTokenizer`, `BridgeEncoding` validation, and `BridgeDataset`. The ragged lists may need to stay for a deprecation window — `Model.phonology_decoder_loop` appends to `generated_phon_tokens` incrementally during generation, so the generation path needs its own incremental padded buffer rather than a rebuild per step.

### Part 2 — `embed_phon_tokens` becomes a gather

```python
def embed_phon_tokens(self, feature_ids: torch.Tensor, feature_mask: torch.Tensor) -> torch.Tensor:
    # feature_ids/(mask): (B, L, F_max)
    embedded = self.phonology_embedding(feature_ids) * feature_mask.unsqueeze(-1)
    averaged = embedded.sum(2) / feature_mask.sum(-1, keepdim=True).clamp(min=1)
    return averaged + self.phon_position_embedding.weight[None, : feature_ids.shape[1]]
```

`clamp(min=1)` guards the all-masked row (division by zero); those rows are zero anyway because the mask zeroed the summands. Note that padded slots gather embedding row 0 and are then zeroed by the mask, so the padding index value is irrelevant to the result — but it **does** matter for the backward pass, where row 0 receives a zero gradient contribution. That is correct but worth a comment.

## Correctness

The transformation is mathematically identical (same sum, same divisor). Verified empirically:

- **CPU: bitwise identical** — `torch.equal(current, vectorized) is True`, max abs diff `0.0`
- **CUDA: max abs diff `2.4e-7`** — float32 reduction-order difference, well below run-to-run nondeterminism
- **Gradients w.r.t. `phonology_embedding.weight`**: max abs diff `9.5e-7` (CPU) / `3.8e-6` (CUDA), `torch.allclose(atol=1e-6)` passes on both

`tests/domain/model/test_embed_phon_tokens.py` pins the current semantics (ragged inputs, variable feature counts, single-feature phonemes, gradient flow) and should keep passing across the rewrite.

## Acceptance criteria

- [ ] `PhonemeTokenizer.encode` emits padded `(B, L, F_max)` ids + mask
- [ ] `embed_phon_tokens` contains no Python loop over batch/position
- [ ] `tests/domain/model/test_embed_phon_tokens.py` passes unchanged (or is updated only for the new signature, with equivalence to the old implementation asserted)
- [ ] Full suite green; outputs match pre-change within `atol=1e-6`
- [ ] Benchmark rerun showing >20× on the isolated forward and a measurable drop in ms/step

## Related findings from the same profiling pass

Separate issues, listed so they aren't lost:

1. **Generation re-embeds the whole prefix each step** (`model.py:523`, `model.py:626`) — `O(L²)` instead of `O(L)`. Compounds with this issue.
2. **`CharacterTokenizer.encode` fills tensors element-by-element** — was ~33× slower than a single `torch.tensor(rows)` build. *Already fixed* in the domain-simplification pass.
3. **`PhonemeTokenizer.encode` re-allocated `torch.arange` inside a nested loop** — *already fixed* (hoisted to `self._feature_range` in `__init__`).
4. **Per-item GPU syncs in the decode loops** (`model.py` `phono_sample`, `sequence_finished[b]` indexing, `torch.tensor(new_tokens[b])` per batch item) — hundreds of device→host round trips per generation.
5. **EOS check rescans the full generated history every step** — `all(any(eos in token for token in tokens) ...)` where `in` on a tensor is a device comparison; `O(B·L²)` syncs.
6. **The lexicon is parsed twice** (train + test `BridgeDataset` each build their own `BridgeTokenizer`): ~1.1 s and ~81 MB duplicated, which also inflates the periodic `gc.collect()` in `TrainingPipeline` to ~187 ms.

## Appendix — reproduction harness

Save as `bench_embed_phon.py` and run `BENCH_DEV=cpu uv run python bench_embed_phon.py` (or `BENCH_DEV=cuda`).

```python
import os, sys, time
sys.path.insert(0, ".")
import torch
torch.set_num_threads(1)
from bridge.utils import device_manager

DEV = os.environ.get("BENCH_DEV", "cpu")
device_manager._device = torch.device(DEV)

from bridge.domain.data import BridgeDataset
from bridge.domain.datamodels import DatasetConfig, ModelConfig, VocabSpec
from bridge.domain.model import Model

ds = BridgeDataset(DatasetConfig(dataset_filepath="tests/domain/model/data/data.csv"))
vocab = VocabSpec.from_tokenizer(ds.tokenizer)
model = Model(ModelConfig(vocab=vocab, seed=1)).to(device_manager.device)
dev = device_manager.device

B = 32
toks = ds[slice(0, B)].phonological.dec_input_ids

def sync():
    if DEV == "cuda":
        torch.cuda.synchronize()

def bench(fn, n=50, warm=5):
    for _ in range(warm):
        fn()
    sync()
    t = time.perf_counter()
    for _ in range(n):
        fn()
    sync()
    return (time.perf_counter() - t) / n * 1000

# Padded representation built ONCE — this is what the tokenizer would emit.
Bn, L = len(toks), len(toks[0])
Fmax = max(len(t) for b in toks for t in b)
flat_idx = torch.zeros((Bn, L, Fmax), dtype=torch.long, device=dev)
flat_mask = torch.zeros((Bn, L, Fmax), dtype=torch.bool, device=dev)
for b, batch in enumerate(toks):
    for i, t in enumerate(batch):
        flat_idx[b, i, : len(t)] = t
        flat_mask[b, i, : len(t)] = True
counts = flat_mask.sum(-1, keepdim=True).clamp(min=1).to(torch.float)

def vectorized():
    e = model.phonology_embedding(flat_idx) * flat_mask.unsqueeze(-1)
    return e.sum(2) / counts + model.phon_position_embedding.weight[None, :L]

cur, vec = model.embed_phon_tokens(toks), vectorized()
print(f"device={DEV} B={Bn} L={L} Fmax={Fmax} python iters/call={Bn * L}")
print("max abs diff:", (cur - vec).abs().max().item(), "| bitwise:", torch.equal(cur, vec))
t_cur, t_vec = bench(lambda: model.embed_phon_tokens(toks)), bench(vectorized)
print(f"forward current {t_cur:.3f} ms | vectorized {t_vec:.3f} ms | x{t_cur / t_vec:.1f}")

def bwd(f):
    def g():
        f().sum().backward()
        model.zero_grad(set_to_none=True)
    return g

t_cb, t_vb = bench(bwd(lambda: model.embed_phon_tokens(toks)), n=20), bench(bwd(vectorized), n=20)
print(f"fwd+bwd current {t_cb:.3f} ms | vectorized {t_vb:.3f} ms | x{t_cb / t_vb:.1f}")

model.zero_grad(set_to_none=True)
model.embed_phon_tokens(toks).sum().backward()
g_cur = model.phonology_embedding.weight.grad.clone()
model.zero_grad(set_to_none=True)
vectorized().sum().backward()
print("grad allclose:", torch.allclose(g_cur, model.phonology_embedding.weight.grad, atol=1e-6))
```
